File: java/SR-RBD/src/ddb/communication/TcpSender.java
/**
 * 
 */
package ddb.communication;

import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;
import java.net.InetSocketAddress;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.BlockingQueue;

import ddb.Logger;
import ddb.msg.Message;

/**
 * <!-- begin-UML-doc --> <!-- end-UML-doc -->
 * 
 * @author Administrator
 * @generated 
 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
 */
public class TcpSender {

	private final static String LOGGING_NAME = "TcpSender";
	private Map<InetSocketAddress, NodeInfo> nodes = new HashMap<InetSocketAddress, NodeInfo>();
	protected BlockingQueue<Message> queue;
	
	
	/**
	 * @generated "Singleton (com.ibm.xtools.patterns.content.gof.creational.singleton.SingletonPattern)"
	 */
	private static TcpSender instance = null;

	/**
	 * @generated "Singleton (com.ibm.xtools.patterns.content.gof.creational.singleton.SingletonPattern)"
	 */
	protected TcpSender() {
		
	}

	/**
	 * @generated "Singleton (com.ibm.xtools.patterns.content.gof.creational.singleton.SingletonPattern)"
	 */
	public static TcpSender getInstance() {
		// begin-user-code
		if(instance == null)
			instance = new TcpSender();
		
		return instance;
		// end-user-code
	}
	
	public void setQueue(BlockingQueue<Message> q)
	{
		queue = q;
	}
	
	public synchronized void removeNode(InetSocketAddress address) {
		if(nodes.remove(address) == null)
		{
			Logger.getInstance().log("Request to remove unexisting node: " + address.toString(), 
					LOGGING_NAME,
					Logger.Level.WARNING);
		}
		
		Logger.getInstance().log("Node disconnected legally " + address.toString(), 
				LOGGING_NAME,
				Logger.Level.INFO);
	}

	public synchronized void addNodeBySocket(InetSocketAddress node, Socket s) 
	{
		if(nodes.get(node) != null)
		{
			Logger.getInstance().log("Request to add already existing node: " + node.toString(), 
					LOGGING_NAME,
					Logger.Level.WARNING);
		}
		
		nodes.put(node, new NodeInfo(s, false));
	}
	
	public synchronized void AddServerNode(InetSocketAddress node, BlockingQueue<Message> storage)
	{
		NodeInfo nodeInfo = nodes.get(node);
		
		if(nodeInfo == null)
		{
			// connect to node
			try {
				Socket socket = new Socket(node.getAddress(), node.getPort());
				
				TcpWorker worker = new TcpWorker(socket, storage);
                new Thread(worker, 
            		"TcpWorker " + socket.getInetAddress() + ":" + socket.getPort()
        		).start();
                
                nodes.put(node, new NodeInfo(socket, true));
				
			} catch (IOException e) {
				Logger.getInstance().log("Failed to add server node: " + node.toString(), 
						LOGGING_NAME,
						Logger.Level.WARNING);
				return;
			}
			
			
		}
		else
		{
			// mark node as server
			nodeInfo.setIsServer(true);
		}
	}
	
	public synchronized int getServerNodesCount()
	{
		int count = 0;
		
		for(NodeInfo node : nodes.values())
			if(node.getIsServer())
				++count;
		
		// add self
		return count + 1;
	}
	
	/**
	 * /* (non-Javadoc) * @see TcpSender#sendToAllNodes(Message message)
	 * 
	 * @generated 
	 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
	 */
	public synchronized void sendToAllServerNodes(Message message) {
		// begin-user-code
		byte[] data; 
		
		// serialize
		data = message.Serialize();
		
		// search for server nodes
		Iterator<Map.Entry<InetSocketAddress, NodeInfo>> it = nodes.entrySet().iterator();

		while(it.hasNext())
		{
			Map.Entry<InetSocketAddress, NodeInfo> entry = it.next();
			NodeInfo node = entry.getValue();
			
			if(node.getIsServer())
			{
				if(!writeToNode(entry.getKey(), node.getSocket(), data))
				{
					it.remove();
				}
			}
		}
		
		// write to self
		try {
			queue.put(message);
		} catch (InterruptedException e) {
			Logger.getInstance().log("InterruptedException Cannot send to self" , 
					LOGGING_NAME,
					Logger.Level.SEVERE);
		}
		// end-user-code
	}

	/**
	 * /* (non-Javadoc) * @see TcpSender#sendToNode(Message message, String to)
	 * 
	 * @generated 
	 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
	 */
	public synchronized void sendToNode(Message message, InetSocketAddress to) {
		// begin-user-code
		byte[] data; 
		
		if(to == null)
		{
			// write to self
			try {
				queue.put(message);
			} catch (InterruptedException e) {
				Logger.getInstance().log("InterruptedException Cannot send to self" , 
						LOGGING_NAME,
						Logger.Level.SEVERE);
			}
			
			return;
		}
		
		
		// serialize
		data = message.Serialize();
		NodeInfo node = nodes.get(to);
		
		if(node == null)
		{
			Logger.getInstance().log("Request to send to unexisting node: " + to.toString(), 
					LOGGING_NAME,
					Logger.Level.WARNING);
		}
		
		if(!writeToNode(to, node.getSocket(),data))
		{
			nodes.remove(to);
		}
		// end-user-code
	}

	/**
	 * @param data
	 *            the data to be sent -- array of byte arrays
	 * @param socket
	 *            an established socket to the target
	 * @throws IOException
	 */
	private boolean writeToNode(InetSocketAddress node, Socket s, byte[] data) {

		try {
			OutputStream sos = s.getOutputStream();
			sos.write(data);
		} catch (IOException e) {
			
			Logger.getInstance().log("Failure to write to node: " + s.getRemoteSocketAddress(), 
					LOGGING_NAME,
					Logger.Level.WARNING);
			
			// remove node from pool
			nodes.remove(node);
			return false;
		}
		
		return true;
	}






}

File: C#/RBD/RBD/Communication/TcpSender.cs
// +

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Sockets;
using System.Net;
using System.Threading;

using RBD.Util;
using RBD.Msg;
using RBD.Communication;

namespace RBD
{
    class TcpSender
    {
        const string LOGGING_NAME = "TcpSender";        
        SortedDictionary<IPEndPoint, NodeInfo> nodes = new SortedDictionary<IPEndPoint, NodeInfo>();
        static TcpSender instance = new TcpSender();
	    protected BlockingQueue<Message> queue;

        protected TcpSender()
        {
        }

        public static TcpSender getInstance()
        {
            return instance;
        }

        public void setQueue(BlockingQueue<Message> q)
        {
            queue = q;
        }

        public void removeNode(IPEndPoint address)
        {
            lock (this)
            {
                if (nodes.Remove(address) == false)
                {
                    Logger.getInstance().log("Request to remove unexisting node: " + address.ToString(),
                                    LOGGING_NAME,
                                    Logger.Level.WARNING);
                }
                Logger.getInstance().log("Node disconnected " + address.ToString(),
                        LOGGING_NAME,
                        Logger.Level.INFO);
            }
        }

        public void addNodeBySocket(IPEndPoint node, Socket s)
        {
            lock (this)
            {
                if (nodes.ContainsKey(node))
                {
                    Logger.getInstance().log("Request to add already existing node: " + node.ToString(),
                                    LOGGING_NAME,
                                    Logger.Level.WARNING);
                }

                nodes.Add(node, new NodeInfo(s, false));
            }
        }

        public void AddServerNode(IPEndPoint node, BlockingQueue<Message> storage) 
        {
            lock (this)
            {
                NodeInfo nodeInfo = nodes[node];

                if (nodeInfo == null)
                {
                    // connect to node
                    IPEndPoint nodeEndPoint;
                    Socket socket;
                    try
                    {
                        nodeEndPoint = new IPEndPoint(node.Address, node.Port);
                        socket = new Socket(AddressFamily.InterNetwork,
                                    SocketType.Stream, ProtocolType.Tcp);
                        socket.Connect(nodeEndPoint);

                        TcpWorker worker = new TcpWorker(socket, storage);
                        new Thread(new ThreadStart(worker.run)).Start();
                        nodes[node] = new NodeInfo(socket, true);
                    }
                    catch (Exception)
                    {
                        Logger.getInstance().log("Failed to add server node: " + node.ToString(),
                                        LOGGING_NAME,
                                        Logger.Level.WARNING);
                        return;
                    }
                    nodes[node] = new NodeInfo(socket, true);
                }
                else
                {
                    // mark node as server                      
                    nodeInfo.setIsServer(true);
                }
            }
        }

        public int getServerNodesCount()
        {
            lock (this)
            {
                int count = 0;

                foreach (NodeInfo node in nodes.Values)
                    if (node.getIsServer())
                        ++count;

                return count;
            }
        }

        public void sendToAllServerNodes(Message message)
        {
            lock (this)
            {
                // begin-user-code                                                                  
                byte[] data;

                // serialize
                data = message.Serialize();

                // search for server nodes
                IDictionaryEnumerator it = nodes.GetEnumerator();
                //Iterator<Map.Entry<InetSocketAddress, NodeInfo>> it = nodes.entrySet().iterator();  // w javie
                while (it.MoveNext())
                {
                    NodeInfo node = (NodeInfo)it.Value;
                    if (node.getIsServer())
                    {
                        if (!writeToNode((IPEndPoint)it.Key, node.getSocket(), data))
                        {
                            // TODO nie czaję tego kawałka -- usuwamy
                            nodes.Remove((IPEndPoint)it.Key);
                            //it.remove();  // w javie
                        }
                    }
                }

                // write to self
                try
                {
                    queue.put(message);
                }
                catch (ThreadInterruptedException e)
                {
                    Logger.getInstance().log("InterruptedException Cannot send to self",
                            LOGGING_NAME,
                            Logger.Level.SEVERE);
                }
            }
            // end-user-code                                                        
        }

        public void sendToNode(Message message, IPEndPoint to)
        {
            lock (this)
            {
                // begin-user-code
                byte[] data;

                if (to == null)
                {
                    // write to self
                    try
                    {
                        queue.put(message);
                    }
                    catch (ThreadInterruptedException)
                    {
                        Logger.getInstance().log("InterruptedException Cannot send to self",
                                LOGGING_NAME,
                                Logger.Level.SEVERE);
                    }

                    return;
                }

                // serialize
                data = message.Serialize();
                NodeInfo node = nodes[to];

                if (node == null)
                {
                    Logger.getInstance().log("Request to send to unexisting node: " + to.ToString(),
                            LOGGING_NAME,
                            Logger.Level.WARNING);
                }

                if (!writeToNode(to, node.getSocket(), data))
                {
                    nodes.Remove(to);
                }
                // end-user-code    
            }
        }

        private bool writeToNode(IPEndPoint node, Socket s, byte[] data)
        {
            try
            {
                s.Send(data);
            }
            catch (System.IO.IOException)
            {
                Logger.getInstance().log("Failure to write to node: " + node.ToString(),
                                LOGGING_NAME,
                                Logger.Level.WARNING);

                // remove node from pool
                nodes.Remove(node);
                return false;
            }
            return true;
        }
    }
}
